worker/paidAction.js

```diff
@@ -0,0 +1,150 @@
+import { paidActions } from '@/api/paidAction'
+import { datePivot } from '@/lib/time'
+import { Prisma } from '@prisma/client'
+import { getInvoice } from 'ln-service'
+
+export async function settleAction ({ data: { invoiceId }, models, lnd, boss }) {
+  let dbInv
+
+  try {
+    dbInv = await models.invoice.findUnique({
+      where: { id: invoiceId }
+    })
+    const invoice = await getInvoice({ id: dbInv.hash, lnd })
+
+    if (!invoice.is_confirmed) {
+      throw new Error('invoice is not confirmed')
+    }
+
+    await models.$transaction(async tx => {
+      // optimistic concurrency control (aborts if invoice is not in PENDING state)
+      await tx.invoice.update({
+        where: { id: dbInv.id, actionState: 'PENDING' },
+        data: {
+          actionState: 'PAID',
+          confirmedAt: new Date(invoice.confirmed_at),
+          confirmedIndex: invoice.confirmed_index,
+          msatsReceived: BigInt(invoice.received_mtokens)
+        }
+      })
```

Comment:
Mhh, so the question here to make sure that `onPaid` is called exactly once is: **Do other transactions see this update before the transactions commits?**

To answer this question, I read the [section about the Read Committed isolation level](https://www.postgresql.org/docs/current/transaction-iso.html#XACT-READ-COMMITTED), the default isolation level of Postgres:

> `UPDATE`, `DELETE`, `SELECT FOR UPDATE`, and `SELECT FOR SHARE` commands behave the same as `SELECT` in terms of searching for target rows: they will only find target rows that were committed as of the command start time. However, such a target row might have already been updated (or deleted or locked) by another concurrent transaction by the time it is found. **In this case, the would-be updater will wait for the first updating transaction to commit or roll back** (if it is still in progress).

This seems to be what we want here. **Other transactions wait until they can see this update.**

So only one transaction should find a row with the pending state, update it, continue and commit. Every other transaction should wait at this statement and then see that there is no longer such a row. Therefore, it should not continue due to [Prisma returning a `RecordNotFound` exception in this case](https://www.prisma.io/docs/orm/reference/prisma-client-reference#return-type-4).

Since this behavior depends on the isolation level, I think it makes sense to explicitly set this transaction level here even though it should be the default.

Reply:
Read Commited is the lowest isolation level in postgres:

> PostgreSQL's Read Uncommitted mode behaves like Read Committed

Still, being explicit will help readability.